tortoise/verifying_tortoise.go

```diff
@@ -194,428 +309,936 @@ func (t *turtle) checkBlockAndGetInputVector(
 }
 
 // Persist saves the current tortoise state to the database
 func (t *turtle) persist() error {
 	return t.bdp.Persist(mesh.TORTOISE, t)
 }
 
-// RecoverVerifyingTortoise retrieve latest saved tortoise from the database
+// RecoverVerifyingTortoise retrieves the latest saved tortoise from the database
 func RecoverVerifyingTortoise(mdb retriever) (interface{}, error) {
 	return mdb.Retrieve(mesh.TORTOISE, &turtle{})
 }
 
-func (t *turtle) processBlock(block *types.Block) error {
+func (t *turtle) processBlock(ctx context.Context, block *types.Block) error {
+	logger := t.logger.WithContext(ctx).WithFields(
+		log.FieldNamed("processing_block_id", block.ID()),
+		log.FieldNamed("processing_block_layer", block.LayerIndex))
+
 	// When a new block arrives, we look up the block it points to in our table,
 	// and add the corresponding vector (multiplied by the block weight) to our own vote-totals vector.
 	// We then add the vote difference vector and the explicit vote vector to our vote-totals vector.
-	t.logger.With().Debug("processing block", block.Fields()...)
-	t.logger.With().Debug("getting baseblock", block.BaseBlock)
+	logger.With().Debug("processing block", block.Fields()...)
+	logger.With().Debug("getting base block", log.FieldNamed("base_block_id", block.BaseBlock))
 
-	base, err := t.bdp.GetBlock(block.BaseBlock)
+	baseBlock, err := t.bdp.GetBlock(block.BaseBlock)
 	if err != nil {
-		return fmt.Errorf("can't find baseblock")
+		return errBaseBlockNotInDatabase
 	}
 
-	t.logger.With().Debug("block supports", types.BlockIdsField(block.BlockHeader.ForDiff))
-	t.logger.With().Debug("checking baseblock", base.Fields()...)
+	logger.With().Debug("block adds support for",
+		log.Int("count", len(block.BlockHeader.ForDiff)),
+		types.BlockIdsField(block.BlockHeader.ForDiff))
+	logger.With().Debug("checking base block", baseBlock.Fields()...)
 
-	layerOpinions, ok := t.BlockOpinionsByLayer[base.LayerIndex]
+	layerOpinions, ok := t.BlockOpinionsByLayer[baseBlock.LayerIndex]
 	if !ok {
-		return fmt.Errorf("baseblock layer not found %v, %v", block.BaseBlock, base.LayerIndex)
+		return fmt.Errorf("%s: %v, %v", errstrBaseBlockLayerMissing, block.BaseBlock, baseBlock.LayerIndex)
 	}
 
-	baseBlockOpinion, ok := layerOpinions[base.ID()]
+	baseBlockOpinion, ok := layerOpinions[baseBlock.ID()]
 	if !ok {
-		return fmt.Errorf("baseblock not found in layer %v, %v", block.BaseBlock, base.LayerIndex)
+		return fmt.Errorf("%s: %v, %v", errstrBaseBlockNotFoundInLayer, block.BaseBlock, baseBlock.LayerIndex)
+	}
+
+	calcWeightedVote := func(block *types.Block, votedOn types.BlockID, baseVote vec) (vec, error) {
+		weight, err := t.voteWeight(ctx, block, votedOn)
+		if err != nil {
+			return baseVote, fmt.Errorf("error processing block %v in block %v for diff list: %w", votedOn, block.ID(), err)
+		}
+		return baseVote.Multiply(weight), nil
 	}
 
+	// TODO: this logic would be simpler if For and Against were a single list
```

Comment:
agreed

Reply:
also a high priority follow up item, in #2369